10_rest/n0npax/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	puppy "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy/store"
+	"github.com/gin-gonic/gin"
+)
+
+var (
+	out io.Writer = os.Stderr
+)
+
+func main() {
+	var puppyFilePath, storeType string
+	var port int
+	kingpin.Flag("data", "path to file with puppies data").Short('d').ExistingFileVar(&puppyFilePath)
+	kingpin.Flag("port", "Port number").Short('p').Default("8181").IntVar(&port)
+	kingpin.Flag("store", "Store type").Short('s').Default("map").EnumVar(&storeType, "map", "sync", "db")
+	kingpin.Parse()
+
+	_, err := kingpin.CommandLine.Parse(os.Args[1:])
+	checkError(err)
+
+	storer := createStorer(storeType)
+
+	if puppyFilePath != "" {
+		puppies := readPuppies(puppyFilePath)
+		feedStore(storer, puppies)
+	}
+	r := puppy.RestBackend(storer)
+	runSrv(r, port)
+}
+
+func runSrv(r *gin.Engine, port int) {
+	err := r.Run(fmt.Sprintf(":%d", port))
+	checkError(err)
+}
+
+func createStorer(s string) puppy.Storer {
+	switch s {
+	case "db":
+		return store.NewLevelDBStorer()
+	case "sync":
+		return store.NewSyncStore()
+	case "map":
+		return store.NewMemStore()
+	default:
+		panic("unknown storer type")
```

Comment:
return error rather than panic. whether or not to panic should usually be left to top level (main). makes it somewhat easier to unit test too.